Ginkgo: Migrated 01-guestbook test

eloycoto opened the pull request:
- Migrated tests/k8s/tests/01-guestbook-test.sh to ginkgo
- Added the same policy in different namespace in a different test.

I didn't deprecated the 01-guestbook mainly because `k8sT/policies` are not validated yet. 

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
Ginkgo: Migrated 01-guestbook test

- Migrated tests/k8s/tests/01-guestbook-test.sh to ginkgo
- Added the same policy in different namespace in a different test.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -200,6 +200,16 @@ func (kub *Kubectl) NodeCleanMetadata() error {
 	return nil
 }
 
+// NamespaceCreate it creates a new kubernetes namespace with the given name
```
remove "it".
kubernetes --> Kubernetes.

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -200,6 +200,16 @@ func (kub *Kubectl) NodeCleanMetadata() error {
+	return kub.Exec(fmt.Sprintf("%s create namespace %s", KubectlCmd, name))
+}
+
+// NamespaceDelete it deletes a given kubernetes namespace
```
remove "it".
kubernetes --> Kubernetes.

ianvernon commented on test/k8sT/Policies.go:
```diff
@@ -88,197 +80,421 @@ var _ = Describe("K8sPolicyTest", func() {
-		ciliumPod, err := kubectl.GetCiliumPodOnNode(helpers.KubeSystemNamespace, helpers.K8s1)
-		Expect(err).Should(BeNil())
+		It("PolicyEnforcement Changes", func() {
+			// This is a small test that check that basic policyEnforcement
```
that check --> that checks

policyEnforcement --> policy enforcement

ianvernon commented on test/k8sT/Policies.go:
```diff
@@ -88,197 +80,421 @@ var _ = Describe("K8sPolicyTest", func() {
+		It("PolicyEnforcement Changes", func() {
+			// This is a small test that check that basic policyEnforcement
+			// changes are working in k8s. All the policy enforcements
+			// changes/combinations are tested on runtime/Policies
```
tested on --> tested in 

runtime/Policies.go

ianvernon commented on test/k8sT/Policies.go:
```diff
@@ -88,197 +80,421 @@ var _ = Describe("K8sPolicyTest", func() {
-		Expect(err).Should(BeNil())
+		It("PolicyEnforcement Changes", func() {
+			// This is a small test that check that basic policyEnforcement
+			// changes are working in k8s. All the policy enforcements
```
policy enforcement

ianvernon commented on test/k8sT/Policies.go:
```diff
@@ -88,197 +80,421 @@ var _ = Describe("K8sPolicyTest", func() {
+			kubectl.Delete(kubectl.ManifestGet(deployment))
+			kubectl.Delete(kubectl.ManifestGet(webPolicy))
+			kubectl.Delete(kubectl.ManifestGet(redisPolicy))
+			kubectl.Delete(kubectl.ManifestGet(redisPolicyDeprecated))
```
Since calls to delete resources in Kubernetes are asynchronous, add checks to ensure that deletion occurred successfully, as we want to ensure that the test only ends when everything is cleaned up.

ianvernon commented on test/k8sT/Policies.go:
```diff
@@ -88,197 +80,421 @@ var _ = Describe("K8sPolicyTest", func() {
+
+			eps = kubectl.CiliumEndpointPolicyVersion(ciliumPod)
+			_, err = kubectl.CiliumPolicyAction(
+				helpers.KubeSystemNamespace, kubectl.ManifestGet(redisPolicyDeprecated),
```
The bash test only imports the deprecated Redis policy if we are running Kubernetes 1.6. Won't we be importing the deprecated policy for each K8s version we use?

eloycoto replied:
I think that the both version should be tested.  Both should work in kubernetes 1.7 and we should be sure that it's tested correctly ;-)

ianvernon replied:
You're right, I actually misread the bash test when looking side-by-side.

eloycoto committed:
Ginkgo: Fix PR comments

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

ianvernon commented on test/k8sT/Policies.go:
```diff
@@ -88,197 +80,425 @@ var _ = Describe("K8sPolicyTest", func() {
+			_, err := kubectl.CiliumPolicyAction(
+				helpers.KubeSystemNamespace, kubectl.ManifestGet(webPolicy),
+				helpers.KubectlApply, 300)
+			Expect(err).Should(BeNil(), "Cannot apply web-policy err: %q", err)
```
using "err" is redundant; use `... : %q"), err)`

ianvernon commented on test/k8sT/Policies.go:
```diff
@@ -88,197 +80,425 @@ var _ = Describe("K8sPolicyTest", func() {
+			_, err = kubectl.CiliumPolicyAction(
+				helpers.KubeSystemNamespace, kubectl.ManifestGet(redisPolicy),
+				helpers.KubectlApply, 300)
+			Expect(err).Should(BeNil(), "Cannot apply redis policy err:%q", err)
```
using "err" is redundant; use `... : %q"), err)`

ianvernon commented on test/k8sT/Policies.go:
```diff
@@ -88,197 +80,425 @@ var _ = Describe("K8sPolicyTest", func() {
+				"Web policy cannot be deleted")
+			kubectl.Delete(kubectl.ManifestGet(redisPolicy)).ExpectSuccess(
+				"Redis policy cannot be deleted")
+			kubectl.Delete(kubectl.ManifestGet(redisPolicyDeprecated)).ExpectSuccess(
```
Please address my prior comment be sure to check for deletion of these resources propagated correctly to Cilium. Something like this is done in `Policies Across Namespaces` which is a test in this same file.

eloycoto replied:
Done!

ianvernon replied:
This is not done. Starting at line 405: https://github.com/cilium/cilium/pull/2648/files#diff-3e28e952665f25f47cc466d4d5c9a84cR405

The delete call will "succeed", but since Kubernetes is asynchronous, we need to wait for the resources that are being deleted to be cleaned up. This ensures that the test only exits once deletion of resources is completed. This provides a guarantee that the test environment has no leftovers from the test that just ran.

aanm commented on test/helpers/cons.go:
```diff
@@ -135,6 +135,9 @@ const (
 	KubectlDelete = ResourceLifeCycleAction("delete")
 	KubectlApply  = ResourceLifeCycleAction("apply")
+
+	KubectlPolicyNameLabel      = "io.cilium.k8s-policy-name"
```
Shouldn't we import this labels directly from cilium?

eloycoto replied:
It's a topic already discussed. If we use the cilium labels, on macOS can't using ginkgo. cc @ianvernon

ianvernon replied:
For more context, see: https://github.com/cilium/cilium/issues/2308

aanm replied:
#2702 addresses this, no?

eloycoto replied:
yes!

aanm commented on test/k8sT/Policies.go:
```diff
@@ -88,197 +80,428 @@ var _ = Describe("K8sPolicyTest", func() {
+
+		AfterEach(func() {
+			kubectl.Delete(demoPath)
+			// TO make sure that are not in place
```
this is a TODO?

eloycoto replied:
nope

aanm commented on test/k8sT/Policies.go:
```diff
@@ -88,197 +80,428 @@ var _ = Describe("K8sPolicyTest", func() {
-		// is in runtime/Policies
-		ciliumPod, err := kubectl.GetCiliumPodOnNode(helpers.KubeSystemNamespace, helpers.K8s1)
-		Expect(err).Should(BeNil())
+		It("PolicyEnforcement Changes", func() {
```
Can you make a little bit more descriptive?

aanm commented on test/k8sT/Policies.go:
```diff
@@ -88,197 +80,428 @@ var _ = Describe("K8sPolicyTest", func() {
+			// changes are working in k8s. All the policy enforcement
+			// changes/combinations are tested in runtime/Policies.go
+			ciliumPod, err := kubectl.GetCiliumPodOnNode(helpers.KubeSystemNamespace, helpers.K8s1)
+			Expect(err).Should(BeNil())
```
Add a description `Should(BeNil(), fmt.Sprintf("Getting cilium pod on node %s", helpers.K8s1))`

aanm commented on test/k8sT/Policies.go:
```diff
@@ -88,197 +80,428 @@ var _ = Describe("K8sPolicyTest", func() {
+			status := kubectl.CiliumExec(
+				ciliumPod, fmt.Sprintf("cilium config %s=%s",
+					helpers.PolicyEnforcement, helpers.PolicyEnforcementDefault))
+			status.ExpectSuccess()
```
Add description in case the success was not reached

ianvernon reviewed (commented):
Everything looks fine except for the checking that deletion of resources has indeed occurred - see my comment here https://github.com/cilium/cilium/pull/2648#discussion_r165182357

aanm commented on test/k8sT/Policies.go:
```diff
@@ -88,197 +80,428 @@ var _ = Describe("K8sPolicyTest", func() {
+				ciliumPod, fmt.Sprintf("cilium config %s=%s",
+					helpers.PolicyEnforcement, helpers.PolicyEnforcementDefault))
+			status.ExpectSuccess()
+			helpers.Sleep(5)
```
Why do we need the sleep?

aanm commented on test/k8sT/Policies.go:
```diff
@@ -88,197 +80,428 @@ var _ = Describe("K8sPolicyTest", func() {
+			Expect(policyStatus[models.EndpointPolicyEnabledBoth]).Should(Equal(0))
+		}, 500)
+
+		It("Policies", func() {
```
Please more descriptive with the `It`

eloycoto committed:
Ginkgo: Fix PR comments

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
Ginkgo: Added kubectl.CiliumIsPolicyLoaded

Add a new helper to check if the policy is loaded correctly or not.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

ianvernon commented on test/k8sT/Policies.go:
```diff
@@ -401,17 +401,29 @@ var _ = Describe("K8sPolicyTest", func() {
 			kubectl.Delete(kubectl.ManifestGet(redisPolicyDeprecated)).ExpectSuccess(
 				"Redis deprecated policy cannot be deleted")
+			kubectl.Delete(kubectl.ManifestGet(deployment)).ExpectSuccess(
+				"Guestbook deployment cannot be deleted")
```
This is halfway there, but still not quite there yet. I will clarify more what I think is needed here. Calls to `kubectl delete...` are asynchronous. The resources are not truly "deleted" after the call to `kubectl delete` returns. We need to wait until the pods are no longer present in the output of `kubectl get pods`, to ensure that the pods are fully cleaned up while this test is still running. I'm cautious about leftover cruft from prior tests interfering with test runs.

eloycoto committed:
Ginkgo: Guestbook wait until pods are deleted

- Add a new function that waits until pods are not deleted correctly
from the Kubernetes cluster

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

ianvernon commented on test/k8sT/Policies.go:
```diff
@@ -407,6 +407,23 @@ var _ = Describe("K8sPolicyTest", func() {
+				return false
+			}
+
+			err := helpers.WithTimeout(body, "Pods do not finished correctly",
```
change to "Pods were not able to be deleted" - "finished" is unclear.

ianvernon commented on test/k8sT/Policies.go:
```diff
@@ -407,6 +407,23 @@ var _ = Describe("K8sPolicyTest", func() {
+
+			err := helpers.WithTimeout(body, "Pods do not finished correctly",
+				&helpers.TimeoutConfig{Timeout: helpers.HelperTimeout})
+			Expect(err).To(BeNil(), "Pods didn't termintate correctly")
```
Typo - "termintate" --> "terminate"

ianvernon reviewed (approved):
Small nitpicks / typos. LGTM apart from those; since they are cosmetic in nature, you don't have to reassign to me for review.
